nexus/src/db/datastore/snapshot.rs

```diff
@@ -0,0 +1,166 @@
+        opctx.authorize(authz::Action::Modify, authz_snapshot).await?;
+
+        use db::schema::snapshot::dsl;
+
+        let next_gen: Generation = old_gen.next().into();
+
+        diesel::update(dsl::snapshot)
+            .filter(dsl::id.eq(authz_snapshot.id()))
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::gen.eq(old_gen))
+            .set((dsl::state.eq(new_state), dsl::gen.eq(next_gen)))
+            .returning(Snapshot::as_returning())
+            .get_result_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::NotFoundByResource(authz_snapshot),
+                )
+            })
+    }
+
+    pub async fn project_list_snapshots(
+        &self,
+        opctx: &OpContext,
+        authz_project: &authz::Project,
+        pagparams: &DataPageParams<'_, Name>,
+    ) -> ListResultVec<Snapshot> {
+        opctx.authorize(authz::Action::ListChildren, authz_project).await?;
+
+        use db::schema::snapshot::dsl;
+        paginated(dsl::snapshot, dsl::name, &pagparams)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::project_id.eq(authz_project.id()))
+            .select(Snapshot::as_select())
+            .load_async::<Snapshot>(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
+    }
+
+    pub async fn project_delete_snapshot(
+        &self,
+        opctx: &OpContext,
+        authz_snapshot: &authz::Snapshot,
+        db_snapshot: &Snapshot,
+    ) -> Result<Uuid, Error> {
+        opctx.authorize(authz::Action::Delete, authz_snapshot).await?;
+
+        use db::schema::snapshot::dsl;
+        let now = Utc::now();
+
+        // A snapshot can be deleted in any state. It's never attached to an
+        // instance, and any disk launched from it will copy and modify the volume
+        // construction request it's based on.
+
+        let snapshot_id = authz_snapshot.id();
+        let gen = db_snapshot.gen;
+        let volume_id = db_snapshot.volume_id;
+
+        diesel::update(dsl::snapshot)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::gen.eq(gen))
+            .filter(dsl::id.eq(snapshot_id))
+            .set(dsl::time_deleted.eq(now))
+            .check_if_exists::<Snapshot>(snapshot_id)
+            .execute_and_check(self.pool())
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::Snapshot,
+                        LookupType::ById(snapshot_id),
+                    ),
+                )
+            })?;
+
+        // After deleting the snapshot, remove the associated volume.
+        self.volume_delete(volume_id).await?;
```

Comment:
Should this be in a transaction? What if we mark the snapshot as deleted, but crash before calling `volume_delete`?

Reply:
done in d2fb15d9